training/tf/tfprocess.py

```diff
@@ -49,81 +51,85 @@ def __init__(self, cfg, next_train_batch, next_test_batch=None, num_eval = 10):
-        self.y_ = next_train_batch[1] # tf.placeholder(tf.float32, [None, 1924])
-        self.z_ = next_train_batch[2] # tf.placeholder(tf.float32, [None, 1])
-        self.training = tf.placeholder(tf.bool)
-        self.batch_norm_count = 0
-
-        with tf.device('/device:GPU:{}'.format(self.cfg['gpu'])):
-            self.y_conv, self.z_conv = self.construct_net(self.x)
-
-            # loss on policy head
-            cross_entropy = \
-                tf.nn.softmax_cross_entropy_with_logits(labels=self.y_,
-                                                        logits=self.y_conv)
-            self.policy_loss = tf.reduce_mean(cross_entropy)
-
-            # Loss on value head
-            self.mse_loss = \
-                tf.reduce_mean(tf.squared_difference(self.z_, self.z_conv))
-
-        # Regularizer
-        regularizer = tf.contrib.layers.l2_regularizer(scale=0.0001)
-        reg_variables = tf.trainable_variables()
-        self.reg_term = \
-            tf.contrib.layers.apply_regularization(regularizer, reg_variables)
-        
-        pol_loss_w = self.cfg['training']['policy_loss_weight']
-        val_loss_w = self.cfg['training']['value_loss_weight']
-        loss = pol_loss_w * self.policy_loss + val_loss_w * self.mse_loss + self.reg_term
-
-        # Following AlphaGo Zero paper for supervised learningparameters
-        starter_learning_rate = self.cfg['training']['learning_rate']
-        decay_step = self.cfg['training']['decay_step']
-        decay_rate = self.cfg['training']['decay_rate']
-        learning_rate = tf.train.exponential_decay(starter_learning_rate, self.global_step, decay_step, decay_rate, staircase=True)
-        opt_op = tf.train.MomentumOptimizer(
-            learning_rate, momentum=0.9, use_nesterov=True)
-
-        self.update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
-        with tf.control_dependencies(self.update_ops):
-            self.train_op = \
-                opt_op.minimize(loss, global_step=self.global_step)
-
-        correct_prediction = \
-            tf.equal(tf.argmax(self.y_conv, 1), tf.argmax(self.y_, 1))
-        correct_prediction = tf.cast(correct_prediction, tf.float32)
-        self.accuracy = tf.reduce_mean(correct_prediction)
+        with tf.device('/cpu:0'):
+            self.next_train_batch = next_train_batch
+            self.next_test_batch = next_test_batch if next_test_batch else next_train_batch
+            self.global_step = tf.Variable(0, name='global_step', trainable=False)
+            self.x = next_train_batch[0]  # tf.placeholder(tf.float32, [None, 120, 8 * 8])
+            self.y_ = next_train_batch[1] # tf.placeholder(tf.float32, [None, 1924])
+            self.z_ = next_train_batch[2] # tf.placeholder(tf.float32, [None, 1])
+            self.training = tf.placeholder(tf.bool)
+            self.batch_norm_count = 0
+
+            with tf.device('/device:GPU:{}'.format(self.cfg['gpu'])):
```

Comment:
What is the effect of nesting the context here?

Reply:
I believe it will map everything it can to the GPU and everything it can not to the cpu. It's based on the example here https://github.com/tensorflow/models/blob/master/tutorials/image/cifar10/cifar10_multi_gpu_train.py. But yeah, I agree it's dodgy.

Reply:
I tried this and it crashed on stuff it could not compute on the gpu. I found that it was easier to choose the used gpu by using `visible_device_list`(https://github.com/tensorflow/tensorflow/blob/r1.4/tensorflow/core/protobuf/config.proto) to set the gpu I wanted to use as the only gpu visible.

Reply:
@zediir The option `tf.ConfigProto(allow_soft_placement=True)` probably helps there. But I agree that your method is better anyway.